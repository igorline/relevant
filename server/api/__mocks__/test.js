// exports.test = 'test';

module.exports = 'test';
