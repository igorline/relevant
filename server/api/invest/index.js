import express from 'express';
import controller from './invest.controller';
import auth from '../../auth/auth.service';

const router = express.Router();

router.post('/bet', auth.isAuthenticated(), auth.communityMember(), controller.bet);

router.post('/', auth.isAuthenticated(), auth.communityMember(), controller.create);
router.get('/downvotes', auth.hasRole('admin'), controller.downvotes);
router.get('/:userId', auth.blocked(), auth.communityMember(), controller.show);
router.get('/post/:postId', auth.authMiddleware(), controller.postInvestments);

module.exports = router;
