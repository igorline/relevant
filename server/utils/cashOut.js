import Earnings from '../api/earnings/earnings.model';

export async function logCashOut(user, cashOutAmt, next) {
  try {
    const earnings = await Earnings.find({
      user: user._id,
      status: 'paidout',
      earned: { $gt: 0 }
    });
    const totalPreviousPaidout = earnings.reduce(getTotalEarnings, 0);
    const cashOutLog = new Earnings({
      user: user._id,
      cashOutAttempt: true,
      cashOutAmt,
      prevBalance: user.balance,
      endBalance: user.balance - cashOutAmt,
      totalPreviousPaidout,
      legacyAirdrop: user.legacyAirdrop,
      referralTokens: user.referralTokens,
      airdropTokens: user.airdropTokens,
      status: 'in progress',
      source: 'wallet'
    });
    await cashOutLog.save();
    return cashOutLog;
  } catch (err) {
    return next(new Error('Error logging cashOut', err));
  }
}

export function getTotalEarnings(total, currentVal) {
  total += currentVal.earned;
  return total;
}
