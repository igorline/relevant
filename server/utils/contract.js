import { Contract, ethers } from 'ethers';

const erc20abi = [
  {
    constant: true,
    inputs: [
      {
        name: '_owner',
        type: 'address'
      }
    ],
    name: 'balanceOf',
    outputs: [
      {
        name: 'balance',
        type: 'uint256'
      }
    ],
    payable: false,
    type: 'function'
  }
];

export const getTokenBalance = async ({ address, tokenAddress }) => {
  const provider = ethers.getDefaultProvider('mainnet');
  const contract = new Contract(tokenAddress, erc20abi, provider);
  const balance = await contract.balanceOf(address);
  return balance.toString();
};
