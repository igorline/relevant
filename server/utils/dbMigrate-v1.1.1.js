import CommunityMember from 'server/api/community/community.member.model';
import Community from 'server/api/community/community.model';

async function markDeletedMembers() {
  const deletedCom = await Community.find({ inactive: true }, '_id');
  const ids = deletedCom.map(c => c._id);
  return CommunityMember.updateMany(
    { communityId: { $in: ids } },
    { deletedCommunity: true },
    { new: true }
  );
}

async function updateDB() {
  try {
    await markDeletedMembers();
  } catch (err) {
    console.log(err); // eslint-disable-line
  }
}

updateDB();
