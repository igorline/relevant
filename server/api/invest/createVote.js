import Earnings from 'server/api/earnings/earnings.model';
import { sendNotification as sendPushNotification } from 'server/notifications';
import computeApproxPageRank from 'server/pagerank/computeApproxPageRank';
import { computePostPayout } from 'app/utils/rewards';
import Community from 'server/api/community/community.model';
import Post from 'server/api/post/post.model';
import User from 'server/api/user/user.model';
import Subscription from 'server/api/subscription/subscription.model';
import Notification from 'server/api/notification/notification.model';
import Invest from 'server/api/invest/invest.model';

const { NODE_ENV } = process.env;

// General problem: a user with some reputation who hasn't upvoted anyone
// adds much more weight to the first few posts he/she upvotes
// TODO: Solution â€” only start counting weights after N upvotes?

// Rate-limit vote weights?

// Weight allocation attack:
// Above also solves this possible attack...
// 1. Build up some degree of relevance (takes time)
// 2. Create N sibyls ahead of time and upvote each one
// 3. Wait until right before the a given post's payout time
// 4. Each sybil upvotes the post and effectively transferring all of the rep weight to the post
// 5. After, delete the sybils, and to restore original weights
// This is mitigated by not deleting links to sybil nodes upon their deletion :)
// Partial solution: prevent undoing vote after payout
export const create = async (req, res, next) => {
  try {
    const { community, communityId } = req.communityMember;
    const communityInstance = await Community.findOne({ _id: communityId });

    const userId = req.user._id;
    const postId = req.body.post;
    let { amount } = req.body;

    // amount should be on a scale of -1 to 1 to prevent abuse;
    amount = Math.max(-1, amount);
    amount = Math.min(1, amount);

    let { post, user, author } = await queryDb({ userId, postId, communityId });

    post = await unhideTwitterComments({ amount, post, communityId, community });
    // TODO create twitter user authors!

    ratelimitVotes({ user });
    if (author) {
      author.relevance = await ensurePagerank({ user: author, communityInstance });
    }
    if (user) {
      user.relevance = await ensurePagerank({ user, communityInstance });
    }

    let vote = await getExistingVote({ user, post, amount, communityId });
    const undoInvest = !!vote;

    if (undoInvest && vote.isManualBet && vote.stakedTokens) {
      throw new Error('You cannot undo a vote once after you bet on a post.');
    }

    vote = undoInvest
      ? await vote.removeVote({ post, user })
      : await Invest.createVote({
          post,
          community,
          communityId,
          communityInstance,
          amount,
          user
        });

    const adjustVotes = undoInvest ? -1 : 1;
    post.data.upVotes += amount > 0 ? adjustVotes : 0;
    post.data.downVotes += amount < 0 ? adjustVotes : 0;
    post.data = await post.data.save();

    const authorPagerank = author && (author.relevance.pagerank || 0);
    const initialPostRank = post.data.pagerank || 0;

    const updatePageRank = await computeApproxPageRank({
      communityId,
      author,
      post,
      vote,
      user,
      undoInvest
    });
    if (updatePageRank) ({ author, post } = updatePageRank);

    if (vote && !undoInvest) {
      vote.rankChange = post.data.pagerank - initialPostRank;
      await vote.save();
    }

    await post.updateRank({ communityId });
    post.data.expectedPayout = computePostPayout(post.data, communityInstance);

    post = await post.save();

    if (post.parentPost) {
      await post.parentPost.updateRank({ communityId });
      await post.parentPost.save();
    }

    const subscription = await updateSubscriptions({
      post,
      user,
      amount,
      communityId,
      community,
      undoInvest
    });

    res.status(200).json({
      investment: vote,
      success: true,
      subscription,
      undoInvest,
      rankChange: post.data.pagerank - initialPostRank
    });

    // TODO - put the rest into queue on worker?
    post.updateClient();

    Earnings.updateEarnings({ post, communityId });

    // updates user votes
    user.voteCount = await Invest.countDocuments({
      investor: user._id,
      amount: { $gt: 0 }
    });

    // updates follow & follower counts - deprecated?
    user = await user.getSubscriptions();
    user = await user.save();

    user.updateClient();

    // updates author reputation and send notificaitons
    await processNotifications({
      author,
      community,
      post,
      user,
      amount,
      authorPagerank,
      undoInvest,
      communityId
    });
  } catch (err) {
    next(err);
  }
};

async function getExistingVote({ user, post, amount, communityId }) {
  const type = amount < 0 ? 'dowvote' : 'upvote';

  // Don't let users downvote posts older than one week
  const now = new Date();
  const oneWeek = 1000 * 60 * 60 * 24 * 7;
  if (
    type === 'downvote' &&
    post.data.postDate < now.getTime() - oneWeek &&
    NODE_ENV === 'production'
  ) {
    throw new Error('you cannot downvote posts older than one week');
  }

  if (user._id.equals(post.user)) {
    throw new Error('You can not ' + type + ' your own comment');
  }

  const vote = await Invest.findOne({
    investor: user._id,
    post: post._id,
    communityId
  });

  if (!vote) return null;

  // TODO undo invest
  const timeElapsed = now.getTime() - new Date(vote.createdAt).getTime();
  // TODO 15m to update post is this fine?
  if (timeElapsed > 15 * 60 * 1000 && NODE_ENV === 'production') {
    throw new Error('You cannot change your vote after 15m');
  }

  if (
    new Date(post.data.payoutTime).getTime() < now.getTime() &&
    NODE_ENV === 'production'
  ) {
    throw new Error('you cannot change your vote after post payout');
  }
  return vote;
}

async function updateSubscriptions({
  post,
  user,
  amount,
  undoInvest,
  communityId,
  community
}) {
  if (amount < 0) return null;
  let subscription = await Subscription.findOne({
    follower: user._id,
    following: post.user,
    communityId
  });
  if (!subscription && undoInvest) return null;
  if (!subscription) {
    subscription = new Subscription({
      follower: user._id,
      following: post.user,
      communityId,
      community,
      amount: 0
    });
  }
  const inc = undoInvest ? Math.max(-3, -subscription.amount) : 3;
  subscription.amount = Math.min(subscription.amount + inc, 20);
  return subscription.save();
}

async function sendAuthorNotification({ author, user, post, type, undoInvest, amount }) {
  if (!author) return null;
  // Remove notification if undo;
  if (undoInvest) {
    await Notification.deleteOne({
      type,
      post: post._id,
      forUser: author._id,
      byUser: user._id
    }).exec();
    return null;
  }

  await Notification.createNotification({
    post: post._id,
    forUser: author._id,
    byUser: user._id,
    amount,
    type
  });

  const action =
    type === 'upvote' ? ' thinks your comment is relevant' : ' upvoted a link you shared';

  const payload = {
    fromUser: user,
    toUser: author,
    post,
    action,
    noteType: type
  };
  const alert = user.name + action;

  try {
    sendPushNotification(author, alert, payload);
  } catch (err) {
    console.log(err); // eslint-disable-line
  }
  return null;
}

async function processNotifications(params) {
  const { post, user, amount, authorPagerank, undoInvest } = params;
  let { author } = params;

  if (!author) {
    if (amount < 0) return null;
    const children = await Post.find({
      parentPost: post._id,
      parentComment: { $exists: false }
    }).populate('user');

    const authors = children.map(child => child.user);
    const notes = authors.map(u =>
      sendAuthorNotification({ author: u, user, post, undoInvest, type: 'upvoteParent' })
    );
    return Promise.all(notes);
  }

  const pageRankChange = author.relevance
    ? author.relevance.pagerank - authorPagerank
    : 0;
  const type = 'upvote';

  author = await author.save();
  author.updateClient(user);

  if (amount < 0) return null;

  await sendAuthorNotification({
    author,
    user,
    post,
    undoInvest,
    amount: pageRankChange,
    type
  });
  return author;
}

function ratelimitVotes({ user }) {
  const now = new Date();
  const elapsedTime = new Date(now).getTime() - new Date(user.lastVote || 0).getTime();
  if (elapsedTime < 5 * 1000 && NODE_ENV === 'production') {
    throw new Error('you cannot up-vote posts more often than 5s');
  }
}

async function unhideTwitterComments({ amount, post, communityId, community }) {
  if (!post) return null;
  if (amount > 0 && post.hidden && post.parentPost) {
    await post.parentPost.insertIntoFeed(communityId, community);
    post.hidden = false;
  }
  post.hidden = false;
  return post;
}

async function ensurePagerank({ user, communityInstance }) {
  if (!user) return null;
  if (user.relevance) return user.relevance;
  return communityInstance.join(user._id);
}

async function queryDb({ userId, postId, communityId }) {
  const post = await Post.findOne({ _id: postId })
    .populate({ path: 'parentPost' })
    .populate({ path: 'data', match: { communityId } });

  const user = await User.findOne(
    { _id: userId },
    'name balance ethAddress image lastVote votePower handle tokenBalance lockedTokens notificationSettings'
  ).populate({
    path: 'relevance',
    match: { communityId }
  });

  const author = await User.findOne(
    { _id: post.user },
    'name handle image balance deviceTokens badge'
  ).populate({
    path: 'relevance',
    match: { communityId }
  });
  return { user, author, post };
}
