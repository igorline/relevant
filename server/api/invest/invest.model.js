/* eslint no-console: 0 */
import mongoose from 'mongoose';
import { EventEmitter } from 'events';
import { VOTE_COST_RATIO, SLOPE, EXPONENT } from 'server/config/globalConstants';
// import Community from 'server/api/community/community.model';
import Earnings from 'server/api/earnings/earnings.model';
import { computePostPayout } from 'app/utils/rewards';

const InvestSchemaEvents = new EventEmitter();
const { Schema } = mongoose;

const TEST_ENV = process.env.NODE_ENV === 'test';

const InvestSchema = new Schema(
  {
    post: { type: Schema.Types.ObjectId, ref: 'Post' },
    investor: { type: Schema.Types.ObjectId, ref: 'User' },
    author: { type: Schema.Types.ObjectId, ref: 'User' },

    ownPost: { type: Boolean, default: false },
    amount: Number,

    community: String,
    communityId: { type: Schema.Types.ObjectId, ref: 'Community' },
    // voteWeight is DEPRECATED same as shares - shares is better
    // TODO convert this in old data
    // voteWeight: { type: Number, default: 0 },
    shares: { type: Number, default: 0 },
    stakedTokens: { type: Number, default: 0 },

    paidOut: { type: Boolean, default: false },
    payoutDate: { type: Date },

    // EVERYTHING BELOW SHOULD BE REMOVED - DEPRECATED

    // vote weight
    relevantPoints: { type: Number, default: 0 },
    rankChange: { type: Number, default: 0 },

    // this info helps us determine how much the
    // investor (or author) has earned from this post
    upvotes: { type: Number, default: 0 },
    downVotes: { type: Number, default: 0 },
    updatePowerInvestor: { type: String, ref: 'User' },
    partialUsers: { type: Number, default: 0 },
    relevance: { type: Number, default: 0 },
    partialRelevance: { type: Number, default: 0 }
  },
  {
    timestamps: true
  }
);

// InvestSchema.index({ community: 1 });
InvestSchema.index({ post: 1 });
InvestSchema.index({ investor: 1 });
InvestSchema.index({ communityId: 1, investor: 1 });
InvestSchema.index({ communityId: 1, investor: 1, createdAt: 1 });
InvestSchema.index({ post: 1, investor: 1, ownPost: 1 });
InvestSchema.index({ post: 1, investor: 1, communityId: 1 });

InvestSchema.statics.events = InvestSchemaEvents;

async function undoInvest({ post, investment, user }) {
  post.data.shares -= investment.shares;
  post.data.balance = Math.max(post.data.balance - investment.stakedTokens, 0);
  const returnTokens = Math.min(user.lockedTokens, investment.stakedTokens);
  user.lockedTokens = (user.lockedTokens || 0) - returnTokens;
  console.log('UNLOCK TOKENS', investment.stakedTokens, user.balance, user.lockedTokens);
  post.data.needsRankUpdate = true;

  post.data.totalShares -= investment.stakedTokens;
  const earning = await Earnings.findOne({ user: user._id, post: post._id });
  if (earning) {
    await earning.remove();
    earning.updateClient({ actionType: 'REMOVE_EARNING' });
  }
  await post.data.save();
  await user.save();
  await post.save();
  await investment.remove();
  return investment;
}

InvestSchema.statics.createVote = async function createVote(props) {
  const { post, relevanceToAdd, community, communityId, amount } = props;
  let { user, investment } = props;

  // user = user.updatePower();
  user = await user.updateBalance();
  const userBalance = user.balance + user.tokenBalance;

  investment = await this.findOne({ investor: user._id, post: post._id });
  if (investment) return undoInvest({ post, investment, user });

  let shares = 0;
  let stakedTokens = userBalance * Math.abs(amount) * VOTE_COST_RATIO;

  // can only invest in top-level posts
  let canInvest = false;
  const now = new Date();

  const leeway = TEST_ENV ? 1000 * 60 : 0;

  if (
    !post.parentPost &&
    user.lockedTokens + stakedTokens <= userBalance &&
    post.data.eligibleForReward &&
    !post.data.paidOut &&
    post.data.payoutTime &&
    new Date(post.data.payoutTime).getTime() + leeway > now.getTime()
  ) {
    canInvest = true;
  }

  if (!canInvest) stakedTokens = 0;
  console.log('canInvest', canInvest);

  // TODO - total tokens staked within a community instead
  const communityInstance = await this.model('Community').findOne({ _id: communityId });
  // userBalance *= communityMember.weight;

  // only compute shares for upvotes
  // for downvotes only track staked tokens
  // we can use totalStaked to rank by stake
  // only stake on top-level posts
  if (canInvest) {
    const { balance, shares: postShares } = post.data;
    const nexp = EXPONENT + 1;
    stakedTokens = userBalance * Math.abs(amount) * VOTE_COST_RATIO;
    shares =
      (((Math.max(balance, 0) + stakedTokens) / SLOPE) * nexp) ** (1 / nexp) -
      (postShares || 0);
    console.log(user.handle, 'got', shares, 'for', stakedTokens, 'staked tokens');
  }

  investment = new (this.model('Invest'))({
    investor: user._id,
    post: post._id,
    author: post.user,
    amount,
    relevantPoints: relevanceToAdd,
    ownPost: post.user === user._id,
    shares,
    stakedTokens,
    community,
    communityId,
    // TODO track parentPost && linkPost/aboutLink?
    // parentPost: post.parentPost,
    // linkPost: post.linkPost,
    payoutDate: post.data.payoutDate,
    paidOut: post.data.paidOut
  });
  await investment.save();

  if (!canInvest) return investment;

  post.data.shares += shares;
  post.data.balance += stakedTokens;
  user.lockedTokens += stakedTokens;
  post.data.totalShares += stakedTokens;
  post.data.expectedPayout = computePostPayout(post.data, communityInstance);
  post.data.needsRankUpdate = true;
  await user.save();
  await post.data.save();

  if (amount > 0) {
    await updateUserEarnings({
      user,
      post,
      shares,
      stakedTokens,
      community,
      communityId
    });
  }

  return investment;
};

async function updateUserEarnings({
  user,
  post,
  shares,
  stakedTokens,
  community,
  communityId
}) {
  const earning = await Earnings.findOneAndUpdate(
    { user: user._id, post: post._id },
    {
      shares,
      stakedTokens,
      community,
      communityId,
      payoutTime: post.data.payoutTime,
      estimatedPostPayout: post.data.expectedPayout,
      totalPostShares: post.data.shares,
      status: 'pending'
    },
    { new: true, upsert: true }
  );
  earning.updateClient({ actionType: 'ADD_EARNING' });
}

module.exports = mongoose.model('Invest', InvestSchema);
