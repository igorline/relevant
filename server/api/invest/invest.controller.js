import Invest from 'server/api/invest/invest.model';
import Post from 'server/api/post/post.model';
import Community from 'server/api/community/community.model';
import { create } from './createVote';

exports.create = create;

exports.bet = async (req, res, next) => {
  try {
    const { user } = req;
    const { stakedTokens, postId: _id } = req.body;
    if (stakedTokens === 0) throw new Error("You don't have enough tokens to bet");
    const { communityId, community } = req.communityMember;
    const communityObj = await Community.findOne({ _id: communityId });
    if (communityObj.betEnabled === false) {
      throw new Error("This community doesn't support betting");
    }
    const post = await Post.findOne({ _id }).populate({
      path: 'data',
      match: { communityId }
    });
    let vote = await Invest.findOne({ post: _id, investor: user._id });
    if (!vote) throw new Error('You must upvote the post first');
    vote = await vote.placeBet({ post, community, communityId, stakedTokens, user });
    return res.status(200).json(vote);
  } catch (err) {
    return next(err);
  }
};

exports.postInvestments = async (req, res, next) => {
  try {
    const { postId } = req.params;
    const limit = parseInt(req.query.limit, 10);
    const skip = parseInt(req.query.skip, 10);
    const { community } = req.query;

    const investments = await Invest.find({ post: postId, amount: { $gt: 0 } })
      .limit(limit)
      .skip(skip)
      .sort({ createdAt: -1 })
      .populate({
        path: 'investor',
        select: 'name image handle',
        populate: {
          path: 'relevance',
          match: { community }
        }
      });

    return res.status(200).json(investments);
  } catch (err) {
    return next(err);
  }
};

exports.postvotes = async (req, res, next) => {
  try {
    const { postId } = req.params;
    const limit = parseInt(req.query.limit, 10);
    const skip = parseInt(req.query.skip, 10);
    const { community } = req.query;

    const votes = await Invest.find({ post: postId, amount: { $gt: 0 } })
      .limit(limit)
      .skip(skip)
      .sort({ createdAt: -1 })
      .populate({
        path: 'investor',
        select: 'name image handle',
        populate: {
          path: 'relevance',
          match: { community }
        }
      });

    return res.status(200).json(votes);
  } catch (err) {
    return next(err);
  }
};

exports.downvotes = async (req, res, next) => {
  const limit = parseInt(req.query.limit, 10) || null;
  const skip = parseInt(req.query.skip, 10) || null;
  let downvotes;
  try {
    downvotes = await Invest.find({ amount: { $lt: 0 } })
      .sort({ createdAt: -1 })
      .skip(skip)
      .limit(limit)
      .populate('post');

    return res.status(200).json(downvotes);
  } catch (err) {
    return next(err);
  }
};

exports.show = async (req, res, next) => {
  try {
    const { communityId } = req.communityMember;
    const { user } = req;

    const blocked = user ? [...user.blocked, ...user.blockedBy] : [];

    const limit = parseInt(req.query.limit, 10);
    const skip = parseInt(req.query.skip, 10);
    const userId = req.params.userId || null;
    const sortQuery = { createdAt: -1 };
    const query = { investor: userId, amount: { $gt: 0 } };

    if (blocked.find(u => u.toString() === userId.toString())) {
      return res.status(200).json({});
    }

    const votes = await Invest.find(query)
      .populate({
        path: 'post',
        populate: [
          {
            path: 'embeddedUser.relevance',
            match: { communityId }
          },
          {
            path: 'data',
            select: 'pagerank'
          },
          {
            path: 'metaPost'
          }
        ]
      })
      .limit(limit)
      .skip(skip)
      .sort(sortQuery);
    return res.status(200).json(votes);
  } catch (err) {
    return next(err);
  }
};
