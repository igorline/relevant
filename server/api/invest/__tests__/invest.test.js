import CommunityMember from 'server/api/community/community.member.model';
import { sanitize, toObject } from 'server/test/utils';
import Invest from 'server/api/invest/invest.model';
import Post from 'server/api/post/post.model';
import { getUsers, getPosts, getCommunities } from 'server/test/seedData';
import computePageRank from 'server/pagerank/pagerankCompute';
import { create, bet } from 'server/api/invest/invest.controller';
import { response } from 'jest-mock-express';

// this will define the database name where the tests are run
process.env.TEST_SUITE = 'invest';

jest.mock('server/utils/ethereum');

describe('ethRewards', () => {
  let { alice, bob, relevant, link1, req, res, communityId, communityMember } = {};

  const next = jest.fn(console.log); // eslint-disable-line
  let voteId;

  beforeAll(async () => {
    ({ relevant } = getCommunities());
    ({ alice, bob } = getUsers());
    ({ link1 } = getPosts());
    communityId = relevant._id;
    communityMember = await CommunityMember.findOne({ user: alice._id, communityId });

    const stakedTokens = alice.balance * 0.1;

    req = {
      user: alice,
      body: { post: { _id: link1._id }, amount: 0.5, stakedTokens, postId: link1._id },
      communityMember
    };
    // need to run this to give inital rank to admin
    await computePageRank({ community: relevant.slug, communityId });
  });

  beforeEach(() => {
    res = response();
    global.console.log = jest.fn(); // hides logs
  });

  describe('Invest', () => {
    test('should create invest without bet', async () => {
      await create(req, res, next);
      let apiRes = toObject(res.json.mock.calls[0][0]);
      voteId = apiRes.investment._id;
      apiRes = sanitize(apiRes, 'rankChange');
      expect(apiRes).toMatchSnapshot();
      expect(apiRes.investment.shares).toBe(0);
      expect(apiRes.investment.stakedTokens).toBe(0);
    });

    test('rank should update', async () => {
      const post = await Post.findOne({ _id: link1._id }).populate({
        path: 'data',
        communityId
      });
      expect(post.data.rank).toBeGreaterThan(0);
    });
  });

  describe('Bet', () => {
    test('should bet on post', async () => {
      await bet(req, res, next);
      const { stakedTokens } = req.body;
      let apiRes = toObject(res.json.mock.calls[0][0]);
      apiRes = sanitize(apiRes, 'rankChange');
      expect(apiRes).toMatchSnapshot();

      expect(apiRes.shares).toBeGreaterThan(0);
      expect(apiRes.stakedTokens).toBe(stakedTokens);
    });
  });

  // Todo test all cases of non-invest

  describe('Undo', () => {
    test('should not be able to undo vote after bet', async () => {
      await create(req, res, next);
      expect(next).toBeCalled();
    });
  });

  describe('AutoBet', () => {
    test('should create invest with bet', async () => {
      bob.notificationSettings.bet.manual = false;
      await bob.save();
      req.user = bob;
      await create(req, res, next);
      const apiRes = toObject(res.json.mock.calls[0][0]);
      voteId = apiRes.investment._id;
      expect(apiRes.investment.shares).not.toBe(0);
      expect(apiRes.investment.stakedTokens).not.toBe(0);
    });

    test('should undo auto-bet', async () => {
      await create(req, res, next);
      let apiRes = toObject(res.json.mock.calls[0][0]);
      apiRes = sanitize(apiRes, 'rankChange');
      expect(apiRes).toMatchSnapshot();

      const removedVote = await Invest.findOne({ _id: voteId });
      expect(removedVote).toBe(null);
    });
  });
});
