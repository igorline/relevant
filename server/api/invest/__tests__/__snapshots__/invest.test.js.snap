// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ethRewards AutoBet should undo auto-bet 1`] = `
Object {
  "investment": Object {
    "__v": 0,
    "_id": "[SANITIZED]",
    "amount": 0.5,
    "community": "relevant",
    "communityId": "c00000000000000000000001",
    "createdAt": "[SANITIZED]",
    "downVotes": 0,
    "investor": "b00000000000000000000000",
    "isManualBet": false,
    "ownPost": false,
    "paidOut": false,
    "partialRelevance": 0,
    "partialUsers": 0,
    "post": "000000000000000000000003",
    "rankChange": "[SANITIZED]",
    "relevance": 0,
    "relevantPoints": 0,
    "shares": 1.951644903827174,
    "stakedTokens": 11.100000000000001,
    "updatedAt": "[SANITIZED]",
    "upvotes": 0,
  },
  "rankChange": "[SANITIZED]",
  "subscription": Object {
    "__v": 0,
    "_id": "[SANITIZED]",
    "amount": 0,
    "category": Array [],
    "community": "relevant",
    "communityId": "c00000000000000000000001",
    "createdAt": "[SANITIZED]",
    "follower": "b00000000000000000000000",
    "updatedAt": "[SANITIZED]",
  },
  "success": true,
  "undoInvest": true,
}
`;

exports[`ethRewards Bet should bet on post 1`] = `
Object {
  "__v": 0,
  "_id": "[SANITIZED]",
  "amount": 0.5,
  "community": "relevant",
  "communityId": "c00000000000000000000001",
  "createdAt": "[SANITIZED]",
  "downVotes": 0,
  "investor": "a00000000000000000000000",
  "isManualBet": true,
  "ownPost": false,
  "paidOut": false,
  "partialRelevance": 0,
  "partialUsers": 0,
  "post": "000000000000000000000003",
  "rankChange": "[SANITIZED]",
  "relevance": 0,
  "relevantPoints": 0,
  "shares": 4.711687595755899,
  "stakedTokens": 11.100000000000001,
  "updatedAt": "[SANITIZED]",
  "upvotes": 0,
}
`;

exports[`ethRewards Invest should create invest without bet 1`] = `
Object {
  "investment": Object {
    "__v": 0,
    "_id": "[SANITIZED]",
    "amount": 0.5,
    "community": "relevant",
    "communityId": "c00000000000000000000001",
    "createdAt": "[SANITIZED]",
    "downVotes": 0,
    "investor": "a00000000000000000000000",
    "isManualBet": true,
    "ownPost": false,
    "paidOut": false,
    "partialRelevance": 0,
    "partialUsers": 0,
    "post": "000000000000000000000003",
    "rankChange": "[SANITIZED]",
    "relevance": 0,
    "relevantPoints": 0,
    "shares": 0,
    "stakedTokens": 0,
    "updatedAt": "[SANITIZED]",
    "upvotes": 0,
  },
  "rankChange": "[SANITIZED]",
  "subscription": Object {
    "__v": 0,
    "_id": "[SANITIZED]",
    "amount": 3,
    "category": Array [],
    "community": "relevant",
    "communityId": "c00000000000000000000001",
    "createdAt": "[SANITIZED]",
    "follower": "a00000000000000000000000",
    "updatedAt": "[SANITIZED]",
  },
  "success": true,
  "undoInvest": false,
}
`;
